### DISCUSSION TITLE
Build dbeaver has some problem

### DISCUSSION THREAD

**Discussion by ccczhouxin** [dbeaver/dbeaver#22184] [DBeaver Development]
I has a problem about eclipse,the wiki is this picture.
![image](https://github.com/dbeaver/dbeaver/assets/43168824/a46f70da-a3e3-411d-b924-89e1e933bca9)
i follow the wiki mkdir the new file still has errors.
![image](https://github.com/dbeaver/dbeaver/assets/43168824/8121cffc-4b28-4491-aa99-c00663f4bf59)

![image](https://github.com/dbeaver/dbeaver/assets/43168824/7f949283-7e99-4cfc-9bcb-be51bb779093)
![image](https://github.com/dbeaver/dbeaver/assets/43168824/a3446711-00e8-460b-82e9-a323734a451f)

**Comment by E1izabeth** [comment]
Please try this workaround https://github.com/dbeaver/dbeaver/wiki/Develop-in-Eclipse#resolving-tycho-packaging-plugin-execution-failure

[This workaround](https://github.com/dbeaver/dbeaver/wiki/Develop-in-Eclipse#conflicting-lifecycle-mapping-metadata) seems to stop working on the latest Eclipse version.

Also to avoid this issue you can try from scratch and say "Resolve all later" in the dialog where Eclipse asks for using tycho connector.

**Comment by chirontt** [comment]
@E1izabeth thanks for the workaround involving the `<pluginExecution>` settings, but I think you've missed including the enclosing `<pluginExecutions>` for them, like:
```xml
  <pluginExecutions>
    <pluginExecution>
  ...
    </pluginExecution>
  </pluginExecutions>
```
But after getting rid of those `Conflicting lifecycle mapping...` errors, I'd get ~60,000 compile errors! This is after using either the `lifecycle-mapping-metadata.xml` file as you suggested (and adding those enclosing `<pluginExecutions>`), or ignoring the 'Setup Maven plugin connectors' prompt during DBeaver project import (with 'Resolve All Later' click); they both work but end up with that huge number of compile errors all the same. This is for a freshly new workspace.

FYI: I'm using Eclipse RCP 2023-12 version as now supported in the `devel` main branch. And I don't want to go back to the 2023-09 version.

Eventually I've managed to fix those ~60,000 compile errors, with a couple of MANIFEST.MF changes for some plugins. But I was surprised that these changes weren't needed with the Eclipse 2023-09 version. I guess they are due to some changes in the new version of Eclipse RCP.

I'll submit my changes in a PR, if anyone else gets same compiler errors with their Eclipse RCP 2023-12 version...

**Reply by E1izabeth** [reply]
I just configured DBeaver dev environment on a new eclipse and haven't meet such errors. Could you describe what were these errors about?

**Comment by chirontt** [comment]
Here are my steps:
- get a fresh Eclipse RCP 2023-12, a fresh (empty) workspace, and a newly-cloned DBeaver repo
- run the Eclipse RCP, select the above empty workspace, then install the following software from the update sites as instructed by [Develop in Eclipse](https://github.com/dbeaver/dbeaver/wiki/Develop-in-Eclipse) wiki article:
```
https://p2.dev.dbeaver.com/eclipse-repo/
http://eclipse-color-theme.github.io/update
```
- after restarting Eclipse, import the DBeaver project from the above freshly-cloned repo, with explicit **avoidance** of installing any 'Maven plugin connectors' (by clicking on 'Resolve All Later' then 'Finish' buttons in the diaog), and I end up with more than 60,000 compile errors:

![image](https://github.com/dbeaver/dbeaver/assets/49886739/2186206b-5e26-4242-a4fc-00117d95a897)

**Reply by E1izabeth** [reply]
Hmm.. I haven't meet this. Did you try to do Maven -> Update Project?
I see that you've found a way to fix it, just wondering.

**Reply by chirontt** [reply]
Maven -> Update Projects... didn't/doesn't help.

After some experimentation, I think I've found the root cause of the compile errors: for a fresh workspace, Eclipse at first uses its embedded JRE to compile the newly-imported DBeaver project, and this somehow causes the `org.jkiss.utils` bundle **not** to be resolved (for the rest of the project) resulting in ~66,000 compile errors; this JRE usage is evident in my previous screenshot:

![image](https://github.com/dbeaver/dbeaver/assets/49886739/455f3c5e-e363-4ae9-a0f9-d4c22832b744)

and can be confirmed in the 'Installed JREs' settings:

![image](https://github.com/dbeaver/dbeaver/assets/49886739/73020ce7-b35b-4af6-9e77-c29b0e0c31cb)

Then I just restart Eclipse (after importing the project), the errors are gone, without any fix mentioned below!

In summary, I think the [Develop in Eclipse](https://github.com/dbeaver/dbeaver/wiki/Develop-in-Eclipse) document need be updated:
- in Step 10, the dialog prompt about 'Setup Maven plugin connectors' should be ignored, by clicking on 'Resolve All Later' then 'Finish' buttons in the dialog; and consequently Step 11 is redundant and should be removed. I think this part has caused the 'conflicting lifecycle mapping' errors since at least Eclipse 2023-09 version. And since this Step 10 is now ignored, the 'Troubleshooting' part involving the `lifecycle-mapping-metadata.xml` file is redundant and can be removed too.
- Eclipse should be restarted, if compile errors appear after importing the project.

**Reply by E1izabeth** [reply]
Thank you for investigations

**Comment by chirontt** [comment]
The fix: after noticing that most of those ~60,000 compile errors are due to the **unresolved** bundle `org.jkiss.utils`; somehow in my workspace I need to add the `Bundle-ClassPath: .` entry to its `MANIFEST.MF`:

![image](https://github.com/dbeaver/dbeaver/assets/49886739/8e721a95-d309-4cf3-b807-243b899d8f31)

and voila, all compile errors are gone!

Actually there used to be another compile error in the `org.jkiss.dbeaver.ext.greenplum` bundle, but you have fixed it with recent commit fe63c40.

**Comment by plotn** [comment]
Hi collegues!
I am trying to do the same things (compile under Eclipse RCP 2023-12), Did workaround specified (for org.eclipse.m2e.core), but got an errors (count = 141) (and cannot solve yet). Any suggestions?

---
Failed to execute mojo org.eclipse.tycho:tycho-packaging-plugin:4.0.4:package-plugin {execution: default-package-plugin} (org.eclipse.tycho:tycho-packaging-plugin:4.0.4:package-plugin:default-package-plugin:package)

org.eclipse.core.runtime.CoreException: Failed to execute mojo org.eclipse.tycho:tycho-packaging-plugin:4.0.4:package-plugin {execution: default-package-plugin}
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeMojo(MavenExecutionContext.java:340)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.lambda$0(MavenExecutionContext.java:291)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:394)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:275)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:290)
	at org.eclipse.m2e.core.project.configurator.MojoExecutionBuildParticipant.build(MojoExecutionBuildParticipant.java:57)
	at org.eclipse.m2e.core.internal.builder.MavenBuilderImpl.lambda$2(MavenBuilderImpl.java:153)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:721)
	at org.eclipse.m2e.core.internal.builder.MavenBuilderImpl.build(MavenBuilderImpl.java:133)
	at org.eclipse.m2e.core.internal.builder.MavenBuilder$1.method(MavenBuilder.java:164)
	at org.eclipse.m2e.core.internal.builder.MavenBuilder$1.method(MavenBuilder.java:1)
	at org.eclipse.m2e.core.internal.builder.MavenBuilder$BuildMethod.lambda$1(MavenBuilder.java:109)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:394)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:228)
	at org.eclipse.m2e.core.internal.builder.MavenBuilder$BuildMethod.lambda$0(MavenBuilder.java:100)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeBare(MavenExecutionContext.java:394)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:275)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.execute(MavenExecutionContext.java:214)
	at org.eclipse.m2e.core.internal.builder.MavenBuilder$BuildMethod.execute(MavenBuilder.java:83)
	at org.eclipse.m2e.core.internal.builder.MavenBuilder.build(MavenBuilder.java:192)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:1079)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:296)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:352)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:441)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:47)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:444)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:555)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:503)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:585)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:207)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:300)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-package-plugin of goal org.eclipse.tycho:tycho-packaging-plugin:4.0.4:package-plugin failed: Project org.jkiss.dbeaver:org.jkiss.dbeaver.cmp.simple:eclipse-plugin:6.0.118-SNAPSHOT does not have an expanded version
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:133)
	at org.eclipse.m2e.core.internal.embedder.MavenExecutionContext.executeMojo(MavenExecutionContext.java:338)
	... 32 more
Caused by: java.lang.IllegalStateException: Project org.jkiss.dbeaver:org.jkiss.dbeaver.cmp.simple:eclipse-plugin:6.0.118-SNAPSHOT does not have an expanded version
	at org.eclipse.tycho.core.osgitools.DefaultReactorProject.getExpandedVersion(DefaultReactorProject.java:247)
	at org.eclipse.tycho.packaging.PackagePluginMojo.getManifest(PackagePluginMojo.java:344)
	at org.eclipse.tycho.packaging.PackagePluginMojo.createPluginJar(PackagePluginMojo.java:279)
	at org.eclipse.tycho.packaging.PackagePluginMojo.execute(PackagePluginMojo.java:184)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:126)
	... 33 more

**Comment by chirontt** [comment]
@plotn I can't reproduce your problems with the Eclipse RPC 2023-12. Here's how I set it up:
- a brand-new Eclipse RPC 2023-12 installation
- a brand-new workspace folder
- a brand-new clone of the DBeaver repo
- then I just follow the steps in [Develop in Eclipse](https://github.com/dbeaver/dbeaver/wiki/Develop-in-Eclipse) wiki article, without problems.

Note that, as discussed above, you can choose to apply `Step 10` in the wiki article or not, during the DBeaver project import:
- if you choose to apply `Step 10`, you'll have to also apply the `lifecycle-mapping-metadata.xml` hack (mentioned at the end of the article) after importing the DBeaver project.
- if you choose **not** to apply Step 10 (and 11) -- and this is my preference -- then you **don't** need any `lifecycle-mapping-metadata.xml` hack afterwards.

I've tried it either ways, but I can't reproduce your problems.

**Comment by plotn** [comment]
I did everything the same, I applied "workaround specified (for org.eclipse.m2e.core)". 
But, you know - project is starts normally, thou eclipse says so many errors - this helped: "Sometimes your plugins may depend on other plug-ins that are not included in the DBeaver.product.launch file. You can check it Debug Configurations->Plug-ins tab with the button "Validate Plug-ins". In the dialogue of the problems of the dependencies appears, you can use the magic button "Add Required Plug-ins". Sometimes it solves all dependencies problems."

**Comment by titou10titou10** [comment]
After trying tons of things, because of tons of setup problems, mostly because of conclict between m2e and tychco plugins I finally found a solution to play with the code in eclipse:

A working solution with eclipse RCP v 2023-12 based on the [procedure from the WiKi](https://github.com/dbeaver/dbeaver/wiki/Develop-in-Eclipse) is:

- be sure the "Tycho Projector Configuration plugin" is NOT installed in eclipse beforehand from previous tries...
    Help > About Eclipse IDE, Installation Details, Installed Software, filter on "tycho": should be empty
- keep the `<project root>/.metadata/.plugins/org.eclipse.m2e.core/lifecycle-mapping-metadata.xml`file EMPTY
- FIRST import the`dbeaver-common` project, 
- then import the `dbeaver`project (it depends on`dbeaver-common`..)
- while eclipse and m2e are processing the projects, when asked to install the "Tycho Configurator" in the "setup maven plugin connectors" popup, "select "resolve all later", 
  ie do NOT install the "Tycho Projector Configuration plugin" when asked when importing projects

and voila..
This is simpler that what is in the WiKi, probably because m2e now do much more things than before and will cause conflicts with the "older" tycho maven configurator/plugins

So compared to the current page in the WiKi, the procedure is shorter/simpler
- steps 8 and 9 are removed
- steps 13, 14 are different
- step 15 comes now before step 10
- step 16 and 18 (and 19?) are removed
- step 17, my very own preference is to have the "project presentation "flat" view, but just my preference..

I would also add another step with a very simple/direct way to start/debug DBeaver

    20 to start DBeaver:
      double click on "dbeaver/product/standalone/DBeaver.product (with "project presentation=hierarchical" view)
                   or "dbeaver/product/community/DBeaver.product ("project presentation=flat" view) to open the "product" editor   
      click on the "Launch an Eclipse application" link in the "Overview" tab
                or "Launch an Eclipse application in Debug mode"
      Next time, use the menus "Run/Run" or "Run/Debug" or "Run/Run configurations...and choose the "Dbeaver product" configuration to run DBeaver

This will create a  viable initial "run configuration". The latest versions of Eclipse RCP have the "include required features and plug-ins automatically while launching" checked by default...
